File: docker/executors/Dockerfile.java
# Java OpenJDK 17 Executor
FROM openjdk:17-alpine

# Security: Create non-root user
RUN adduser -D -s /bin/sh coderunner

# Install essential packages
RUN apk add --no-cache \
    gcc \
    musl-dev

# Set working directory
WORKDIR /app

# Copy execution script
COPY run_java.sh /app/
COPY JavaExecutorSimple.java /app/
RUN chmod +x /app/run_java.sh

# Compile the Java executor
RUN javac JavaExecutorSimple.java

# Security configurations
RUN chown -R coderunner:coderunner /app

# Switch to non-root user
USER coderunner

# Set resource limits
ENV EXECUTION_TIMEOUT=10
ENV MEMORY_LIMIT_MB=256

CMD ["/app/run_java.sh"]

File: docker/executors/Dockerfile.cpp
# C++ Executor Dockerfile
FROM gcc:12-bullseye

# Create non-root user
RUN groupadd -r coderunner && useradd -r -g coderunner coderunner

# Install security tools and remove network tools
RUN apt-get update && rm -rf /var/lib/apt/lists/*

# Create app directory
WORKDIR /app

# Copy execution script
COPY run_cpp.sh /app/run_cpp.sh
RUN chmod +x /app/run_cpp.sh

# Set permissions
RUN chown -R coderunner:coderunner /app
RUN chmod 755 /app

# Switch to non-root user
USER coderunner

# Set resource limits and security
ENTRYPOINT ["/app/run_cpp.sh"]
